File: src/RecipeApp.CoreApi/Features/HealthChecks/DatabaseHealthCheck.cs
namespace RecipeApp.CoreApi.Features.HealthChecks;

/// <summary>
/// https://www.youtube.com/watch?v=p2faw9DCSsY
/// Note: this is just an example of custom health check, but is not used in favor of AspNetCore.HealthChecks.SqlServer
/// </summary>
public class DatabaseHealthCheck : IHealthCheck
{
    private readonly DatabaseHealthCheckRepository _databaseHealthCheckRepository;

    public DatabaseHealthCheck(DatabaseHealthCheckRepository databaseHealthCheckRepository) =>
        _databaseHealthCheckRepository = databaseHealthCheckRepository;

    public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
    {
        try
        {
            await _databaseHealthCheckRepository.CheckDatabaseAsync(cancellationToken);

            return HealthCheckResult.Healthy("DatabaseHealthCheck is healthy.");
        }
        catch (Exception ex)
        {
            return HealthCheckResult.Unhealthy("DatabaseHealthCheck is Unhealthy.", ex);
        }
    }
}

File: src/RecipeApp.FCoreApi/Features/HealthChecks/DatabaseHealthCheck.fs
namespace RecipeApp.FCoreApi.Features.HealthChecks

open System.Threading

open Microsoft.Extensions.Diagnostics.HealthChecks

open Tbd.WebApi.Shared.Repositories

type DatabaseHealthCheck(databaseHealthCheckRepository: DatabaseHealthCheckRepository) =
    interface IHealthCheck with
        member this.CheckHealthAsync(context: HealthCheckContext , cancellationToken: CancellationToken) =
            async {

                try
                    do! databaseHealthCheckRepository.CheckDatabaseAsync(cancellationToken) |> Async.AwaitTask

                    return HealthCheckResult.Healthy("DatabaseHealthCheck is healthy.")
                with
                    | ex -> return HealthCheckResult.Unhealthy("DatabaseHealthCheck is Unhealthy.", ex)
            }
            |> Async.StartAsTask
